presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/OptimizeOutputDuplicatesInsensitiveJoinRule.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.Session;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.planner.iterative.GroupReference;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.PlanVisitor;
+import io.prestosql.sql.planner.plan.ProjectNode;
+
+import java.util.Optional;
+
+import static io.prestosql.SystemSessionProperties.isOptimizeOutputDuplicatesInsensitiveJoins;
+import static io.prestosql.sql.planner.DeterminismEvaluator.isDeterministic;
+import static io.prestosql.sql.planner.plan.Patterns.aggregation;
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static java.util.Objects.requireNonNull;
+
+public class OptimizeOutputDuplicatesInsensitiveJoinRule
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(aggregation -> aggregation.getAggregations().isEmpty());
+
+    private final Metadata metadata;
+
+    public OptimizeOutputDuplicatesInsensitiveJoinRule(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeOutputDuplicatesInsensitiveJoins(session);
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregation, Captures captures, Context context)
+    {
+        return aggregation.getSource().accept(new Rewriter(metadata, context.getLookup()), null)
```

Comment:
I agree. In this case, there should be a way to tell optimizer that I want an input plan node which is producing given trait.

We had some ideas around how this can be done in: https://docs.google.com/presentation/d/1rbtJLG89GxEYLZYA09TH5cD4S6UvYXtq0OmdNqQnmws/edit#slide=id.g1fce08ab98_0_4